go/osagent/osagent.go

```diff
@@ -645,3 +745,609 @@ func AbortSeed(seedId string) error {
 func ExecCustomCmdWithOutput(commandKey string) ([]byte, error) {
 	return commandOutput(config.Config.CustomCommands[commandKey])
 }
+
+func contains(item string, list []string) bool {
+	for _, b := range list {
+		if item == b {
+			return true
+		}
+	}
+	return false
+}
+
+// These magical multiplies for logicalSize (0.6 in case of compression, 0.8 in other cases) are just raw estimates. They can be wrong, but we will use them
+// as 'we should have at least' space check, because we can't make any accurate estimations for logical backups
+func GetMySQLDatabaseInfo() (dbinfo MySQLDatabaseInfo, err error) {
+	dbi := make(map[string]*MySQLDatabase)
+	var physicalSize, tokuPhysicalSize, logicalSize int64 = 0, 0, 0
+	databases, err := dbagent.GetMySQLDatabases()
+	for _, db := range databases {
+		engines, err := dbagent.GetMySQLEngines(db)
+		if err != nil {
+			log.Errore(err)
+		}
+		for _, engine := range engines {
+			if engine != "TokuDB" {
+				physicalSize, err = DirectorySize(path.Join(config.Config.MySQLDataDir, db))
+				if err != nil {
+					log.Errore(err)
+				}
+			} else {
+				tokuPhysicalSize, err = dbagent.GetTokuDBSize(db)
+				if err != nil {
+					log.Errore(err)
+				}
+			}
+		}
+		physicalSize += tokuPhysicalSize
+		if config.Config.CompressLogicalBackup {
+			logicalSize = int64(float64(physicalSize) * 0.6)
+		} else {
+			logicalSize = int64(float64(physicalSize) * 0.8)
+		}
+		dbi[db] = &MySQLDatabase{engines, physicalSize, logicalSize}
+		dbinfo.MySQLDatabases = dbi
+	}
+	dbinfo.InnoDBLogSize, err = dbagent.GetInnoDBLogSize()
+	if err != nil {
+		log.Errore(err)
+	}
+	return dbinfo, err
+}
+
+func GetAvailableSeedMethods() []string {
+	var avaliableSeedMethods []string
+	var cmd string
+	for _, seedMethod := range seedMethods {
+		switch seedMethod {
+		case "lvm":
+			cmd = "lvs"
+		case "xtrabackup-stream":
+			cmd = "xtrabackup"
+		default:
+			cmd = seedMethod
+		}
+		err := commandRun(
+			fmt.Sprintf("%s --version", cmd),
+			func(cmd *exec.Cmd) {
+				log.Debug("Checking for seed method", seedMethod)
+			})
+		if err == nil {
+			log.Debug("seed method", seedMethod, "found")
+			avaliableSeedMethods = append(avaliableSeedMethods, seedMethod)
+		}
+	}
+	return avaliableSeedMethods
+}
+
+func StartLocalBackup(seedId string, seedMethod string, databases string) (BackupFolder string, err error) {
+	var cmd string
+	BackupFolder = path.Join(config.Config.MySQLBackupDir, time.Now().Format("20060102-150405"))
+	if err := os.Mkdir(BackupFolder, 0755); err != nil {
+		return "", log.Errore(err)
+	}
+	if !contains(seedMethod, seedMethods) {
```

Comment:
Run this check first, before creating the directory?

Reply:
fixed in 3ce673a